src/ui.cpp

```diff
@@ -696,8 +696,21 @@ void UpdateDialog::OnRunInstaller(wxCommandEvent&)
     }
 }
 
+static inline bool ends_with(std::wstring const& value, std::wstring const& ending)
+{
+    if (ending.size() > value.size()) return false;
+    return std::equal(ending.rbegin(), ending.rend(), value.rbegin());
+}
+
 bool UpdateDialog::RunInstaller()
 {
+    // Jsut let the app deal with it if it's a zip archive.
+    std::wstring _updateFile = m_updateFile;
+    if (ends_with(_updateFile, L".zip") && ApplicationController::_win_sparkle_zip_download_callback) {
+        ApplicationController::_win_sparkle_zip_download_callback(_updateFile);
+        return true;
```

Comment:
And what if it fails to apply?

Reply:
That is true. In terms of winsparkle concepts, do you think a useful alternative would be to allow the app to call a callback when the download payload is complete instead of the default WinSparkle behavior, regardless of file type? Makes sense if you want WinSparkle to keep the entire update workflow in WinSparkle, but how many use cases are unaccounted for? Maybe zip files containing a single aren't the only one? I am ignorant on this issue, but my case is one unaccounted for.

Also, there are enough classes and components in WinSparkle that it took too much energy to grok it all to determine the proper place for various functions. For example, I am not even clear that ```RunInstaller()``` is the correct place to fully handle a download payload like a zip file by: 1) move the running exe to the zip's temp folder (because you can't delete a running exe on windows), 2) unzip the new app exe from the downloaded zip file, and 3) move the new app exe in place, 4) launch the new app exe with ```CreateProcess(**)``` and exit the current process with ```ExitProcess(0)```. If you can suggest a cleaner way to handle this that doesn't break WinSparkle's design principles I can have a look at it.

Reply:
> That is true. In terms of winsparkle concepts, do you think a useful alternative would be to allow the app to call a callback when the download payload is complete instead of the default WinSparkle behavior, regardless of file type? 

Phrased like this, it makes sense to me — thanks for the insight! I was fixating on the ZIP case, but yes, a general callback mechanism to override download's handling is a good idea. 

> I am ignorant on this issue, but my case is one unaccounted for.

I don't feel the need to support *every* crazy case, but your one is reasonable, and it would be nice to support it properly. But until then, having the callback you propose is the pragmatic choice.

> I am not even clear that RunInstaller() is the correct place to fully handle a download payload like a zip file

It probably calls for putting it into a new class/file — as you say, a lot of logic is going to end up there.

> 1) move the running exe to the zip's temp folder (because you can't delete a running exe on windows),

It also needs to move all DLLs from the destination folder (or determine which ones are used by the current process) for the same reason, and it can't move them to the temp folder (permissions; inability to move to another drive). Which only supports your point that it is quite complicated for `RunInstaller()`.

> 4) launch the new app exe with CreateProcess(**) and exit the current process with ExitProcess(0). 

There's a callback for the latter in WinSparkle already (it may be more complicated and must e.g. ensure that files are saved and the user asked about it), so this too can get a bit more complicated.

> If you can suggest a cleaner way to handle this that doesn't break WinSparkle's design principles I can have a look at it.

I think it really needs separate class for all this (presumably specialized for different kinds - installers, ZIPs, delta updates eventually). 

But having an official, documented generic callback to apply the update, which would be used instead of the current launch code if present, is I think the best first step — it has the flexibility to handle unforeseen cases, is enough for your needs for now (if I understand it correctly) *and* you already mostly did it. 

TIA!

Reply:
Very good on all points. I'll start over to produce a viable pull request that simply adds a callback alternative in RunInstaller(). I say "alternative" because you do not want to execute a non-installer payload.